src/App.tsx

```
@@ -8,53 +8,151 @@ import { PostsList } from './components/PostsList';
-          <div className="tile is-child box is-success">
-            <div className="block">
-              <UserSelector />
-            </div>
+export const App = () => {
+  const [users, setUsers] = useState<User[]>([]);
+  const [selectedUser, setSelectedUser] = useState<User | null>(null);
+  const [userPosts, setUserPosts] = useState<Post[]>([]);
+  const [postsLoadingError, setPostsLoadingError] = useState(false);
+  const [loading, setLoading] = useState(false);
+  const [post, setPost] = useState<Post | null>(null);
+  const [comments, setComments] = useState<Comment[]>([]);
+  const [commentLoading, setCommentLoading] = useState(false);
+  const [commentLoadingError, setCommentLoadingError] = useState(false);
+  const [submitLoading, setSubmitLoading] = useState(false);
 
-            <div className="block" data-cy="MainContent">
-              <p data-cy="NoSelectedUser">No user selected</p>
+  useEffect(() => {
+    getUsers().then(u => setUsers(u));
+  }, []);
 
-              <Loader />
+  const handleSelectUser = (user: User) => {
+    setPost(null);
+    setSelectedUser(user);
 
-              <div
-                className="notification is-danger"
-                data-cy="PostsLoadingError"
-              >
-                Something went wrong!
-              </div>
+    setLoading(true);
+    getUserPosts(user.id)
+      .then(posts => setUserPosts(posts))
+      .catch(() => setPostsLoadingError(true))
+      .finally(() => setLoading(false));
+  };
+
+  const handleComments = (p: Post) => {
+    setCommentLoading(true);
+
+    getComments(p.id)
+      .then(c => setComments(c))
+      .catch(() => setCommentLoadingError(true))
+      .finally(() => setCommentLoading(false));
+  };
+
+  const handleDeleteComments = (comment: Comment) => {
+    setComments(currentComments =>
+      currentComments.filter(com => com.id !== comment.id),
+    );
+
+    deleteComments(comment.id).catch(() => setComments(comments));
```

You remove the comment optimistically (good). However the rollback in the catch uses `setComments(comments)` which references the `comments` variable from the closure and can be stale. The task description requires: "Handle `Add` and `Delete` errors so the user can retry" and the checklist recommends using try/catch for error handling. Capture the previous comments (e.g. `const prev = comments`) before removing and restore `prev` in the catch (or use a functional updater). Also surface an error/ retry option to the user instead of silently restoring state.